File: Exemplos/Ex1_Pagamentos/Ex1.js
// PRODUTOS ABSTRATOS
class GatewayPagamento {
    autorizar(valor) { throw new Error("Método abstrato"); }
}

class Recibo {
    gerar(valor) { throw new Error("Método abstrato"); }
}

// PRODUTOS CONCRETOS - PAYPAL
class GatewayPayPal extends GatewayPagamento {
    autorizar(valor) { return `Paypal: Pagamento de R$ ${valor.toFixed(2)} autorizado.`; }
}

class ReciboPayPal extends Recibo {
    gerar(valor) { return `PayPal: Recibo do pagamento de R$ ${valor.toFixed(2)}.`; }
}

// PRODUTOS CONCRETOS - MERCADOPAGO
class GatewayMercadoPago extends GatewayPagamento {
    autorizar(valor) { return `Mercado Pago: Pagamento de R$ ${valor.toFixed(2)} autorizado.`; }
}

class ReciboMercadoPago extends Recibo {
    gerar(valor) { return `Mercado Pago: Recibo do pagamento de R$ ${valor.toFixed(2)}.`; }
}

// FABRICA ABSTRATA
class FabricaPagamento {
    criarGateway() { throw new Error("Método abstrato"); }
    criarRecibo() { throw new Error("Método abstrato"); }
}

// FABRICAS CONCRETAS
class FabricaPayPal extends FabricaPagamento {
    criarGateway() { return new GatewayPayPal(); }
    criarRecibo() { return new ReciboPayPal(); }
}

class FabricaMercadoPago extends FabricaPagamento {
    criarGateway() { return new GatewayMercadoPago(); }
    criarRecibo() { return new ReciboMercadoPago(); }
}

// CLIENTE
function finalizarCompra(fabrica, valor) {
    const gateway = fabrica.criarGateway();
    const recibo = fabrica.criarRecibo();

    console.log(gateway.autorizar(valor));
    console.log(recibo.gerar(valor));
}

finalizarCompra(new FabricaPayPal(), 100.99);
console.log("*************************");
finalizarCompra(new FabricaMercadoPago(), 3100.99);

File: Abstract Factory/Exemplos/Ex1_Pagamentos/Ex1.py
"""
PADRÃO: ABSTRACT FACTORY (FÁBRICA ABSTRATA)

O padrão Abstract Factory é útil quando o sistema deve ser independente de como 
seus produtos são criados, compostos e representados. Ele garante que os produtos 
de uma mesma família sejam usados juntos.
"""

from abc import ABC, abstractmethod

# 1. PRODUTOS ABSTRATOS
class GatewayPagamento(ABC):
    @abstractmethod
    def autorizar(self, valor):
        pass

class Recibo(ABC):
    @abstractmethod
    def gerar(self, valor):
        pass

# 2. PRODUTOS CONCRETOS - FAMÍLIA PAYPAL
class GatewayPayPal(GatewayPagamento):
    def autorizar(self, valor):
        return f"PayPal: Pagamento de R$ {valor:.2f} autorizado"

class ReciboPayPal(Recibo):
    def gerar(self, valor):
        return f"PayPal: Recibo do pagamento de R$ {valor:.2f}."

# 2. PRODUTOS CONCRETOS - FAMÍLIA MERCADOPAGO
class GatewayMercadoPago(GatewayPagamento):
    def autorizar(self, valor):
        return f"Mercado pago: Pagagamento de R$ {valor:.2f} autorizado"

class ReciboMercadoPago(Recibo):
    def gerar(self, valor):
        return f"Mercado Pago: Recibo do pagamento de R$ {valor:.2f}."

# 3. FABRICA ABSTRATA
class FabricaPagamento(ABC):
    @abstractmethod
    def criarGateway(self):
        pass

    @abstractmethod
    def criarRecibo(self):
        pass

# 4. FABRICA CONCRETA
class FabricaPayPal(FabricaPagamento):
    def criarGateway(self):
        return GatewayPayPal()

    def criarRecibo(self):
        return ReciboPayPal()

class FabricaMercadoPago(FabricaPagamento):
    def criarGateway(self):
        return GatewayMercadoPago()

    def criarRecibo(self):
        return ReciboMercadoPago()

# 5. CLIENTE
def finalizarCompra(fabrica: FabricaPagamento, valor: float):
    # O cliente interage com as abstrações (FabricaPagamento)
    gateway = fabrica.criarGateway()
    recibo = fabrica.criarRecibo()

    print(gateway.autorizar(valor))
    print(recibo.gerar(valor))

# Execução
if __name__ == "__main__":
    finalizarCompra(FabricaPayPal(), 100.99)
    print("*************************")
    finalizarCompra(FabricaMercadoPago(), 3100.99)
